sw/device/tests/sram_ctrl_subword_access_test.c

```diff
@@ -0,0 +1,175 @@
+                                uint32_t len) {
+  for (uint32_t i = 0; i < len; ++i) {
+    uint8_t expected = ((uint8_t *)kRandomData)[offset + i];
+    uint8_t got = mmio_region_read8(region, (ptrdiff_t)(offset + i));
+    CHECK(expected == got, "byte %d did not match: expected %02x, got %02x",
+          offset + i, expected, got);
+  }
+}
+
+/// Write each of a u8, u16, u32, and u64 from to every possible byte offset in
+/// SRAM, reading them back to check the writes.
+static void write_subwords_check(mmio_region_t region) {
+  // Check byte writes.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES; ++i) {
+    clear_sram_region(region);
+    ((volatile uint8_t *)region.base)[i] = ((uint8_t *)kRandomData)[i];
+    check_sram_contents(region, i, 1);
+  }
+
+  // Check uint16_t writes.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES - 1; ++i) {
+    clear_sram_region(region);
+    *(volatile uint16_t *)((uint8_t *)region.base + i) =
+        *(uint16_t *)((uint8_t *)kRandomData + i);
+    check_sram_contents(region, i, 2);
+  }
+
+  // Check uint32_t writes.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES - 3; ++i) {
+    clear_sram_region(region);
+    *(volatile uint32_t *)((uint8_t *)region.base + i) =
+        *(uint32_t *)((uint8_t *)kRandomData + i);
+    check_sram_contents(region, i, 4);
+  }
+
+  // Check uint64_t writes.
+  for (uint32_t i = 0; i < SRAM_CTRL_TEST_DATA_SIZE_BYTES - 7; ++i) {
+    clear_sram_region(region);
+    *(volatile uint64_t *)((uint8_t *)region.base + i) =
+        *(uint64_t *)((uint8_t *)kRandomData + i);
+    check_sram_contents(region, i, 8);
+  }
+}
+
+bool test_main(void) {
+  // Initialize SRAM_CTRL hardware.
+  dif_sram_ctrl_t sram_ctrl_main;
+  dif_sram_ctrl_t sram_ctrl_ret;
+  CHECK_DIF_OK(dif_sram_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_SRAM_CTRL_MAIN_REGS_BASE_ADDR),
+      &sram_ctrl_main));
+  CHECK_DIF_OK(dif_sram_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_SRAM_CTRL_RET_AON_REGS_BASE_ADDR),
+      &sram_ctrl_ret));
+
+  dif_sram_ctrl_status_bitfield_t status_main;
+  dif_sram_ctrl_status_bitfield_t status_ret;
+
+  // Check Status registers
+  CHECK_DIF_OK(dif_sram_ctrl_get_status(&sram_ctrl_ret, &status_ret));
+  CHECK_DIF_OK(dif_sram_ctrl_get_status(&sram_ctrl_main, &status_main));
+
+  CHECK((status_main & kStatusRegMask) == 0x0,
+        "SRAM main status error bits set, status = %08x.", status_main);
+  CHECK((status_ret & kStatusRegMask) == 0x0,
+        "SRAM ret status error bits set, status = %08x.", status_ret);
+
+  // Read and Write to/from SRAMs. Main SRAM will use the address of the
+  // buffer that has been allocated. Ret SRAM can start at the base address.
+  uintptr_t sram_main_buffer_addr = (uintptr_t)&sram_main_buffer;
+
+  mmio_region_t sram_region_main_addr =
+      mmio_region_from_addr(sram_main_buffer_addr);
+  mmio_region_t sram_region_ret_base_addr =
```

Comment:
The retention SRAM is used by ROM_EXT, as this test will be booted by ROM_EXT you should mind if this address in not in use. 
See this [example](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/tests/sram_ctrl_sleep_sram_ret_contents_test.c;drc=e79727f409603241d17dfc576182ee12c0e9a25a;l=29
)

Reply:
Oh thanks, I didn't know about that.

This file uses this `.data` section trick for reserving space in the main SRAM, can the same trick be used in retention SRAM? i.e. is there a section that we can place a buffer in that will be reserved there?

Otherwise I'll use the linked method.

Reply:
Actually don't worry, I've used this owner method.

Reply:
The OTTF linker script does not describe any section in the `ram_ret_aon` memory space, but I don't see why you can't add one, specially if you plan to use it in other tests.

Reply:
Okay, I've used the `owner` offset method for the three tests so far, but I'll consider doing this for a follow-up PR since it might be generally useful for owner software using our linker scripts anyway